src/BuildPrediction/Predictors/ArtifactsSdkPredictor.cs

```diff
@@ -0,0 +1,381 @@
+
+        internal const string ArtifactsItemName = "Artifacts";
+
+        internal const string RobocopyItemName = "Robocopy";
+
+        internal const string DestinationFolderMetadata = "DestinationFolder";
+
+        internal const string IsRecursiveMetadata = "IsRecursive";
+
+        internal const string FileMatchMetadata = "FileMatch";
+
+        internal const string FileExcludeMetadata = "FileExclude";
+
+        internal const string DirExcludeMetadata = "DirExclude";
+
+        private static readonly char[] MultiSplits = { '\t', ' ', '\n', '\r' };
+
+        private static readonly char[] Wildcards = { '?', '*' };
+
+        private static readonly char[] DestinationFolderSeparator = { ';' };
+
+        private static readonly char[] DirectorySeparatorChars = { Path.DirectorySeparatorChar };
+
+        /// <inheritdoc/>
+        public void PredictInputsAndOutputs(Project project, ProjectInstance projectInstance, ProjectPredictionReporter predictionReporter)
+        {
+            // This predictor only applies to projects using the Microsoft.Build.Artifacts Sdk.
+            var usingMicrosoftArtifactsSdk = projectInstance.GetPropertyValue(UsingMicrosoftArtifactsSdkPropertyName);
+            if (!usingMicrosoftArtifactsSdk.Equals("true", StringComparison.OrdinalIgnoreCase))
+            {
+                return;
+            }
+
+            // The Sdk allows both Artifacts and Robocopy items
+            ProcessItems(projectInstance.GetItems(ArtifactsItemName), predictionReporter);
+            ProcessItems(projectInstance.GetItems(RobocopyItemName), predictionReporter);
+        }
+
+        private static void ProcessItems(ICollection<ProjectItemInstance> items, ProjectPredictionReporter predictionReporter)
+        {
+            foreach (ProjectItemInstance item in items)
+            {
+                string source = item.EvaluatedInclude;
+                string[] destinationFolders = item.GetMetadataValue(DestinationFolderMetadata).Split(DestinationFolderSeparator, StringSplitOptions.RemoveEmptyEntries);
+                if (string.IsNullOrEmpty(source)
+                    || destinationFolders.Length == 0)
+                {
+                    // Ignore invalid items.
+                    continue;
+                }
+
+                // Make absolute
+                source = Path.GetFullPath(Path.Combine(item.Project.Directory, source)).TrimEnd(DirectorySeparatorChars);
+
+                if (File.Exists(source))
+                {
+                    // If the source is a known file, we can easily predict the inputs and outputs
+                    ProcessFile(source, Path.GetFileName(source), destinationFolders, null, predictionReporter);
+                }
+                else if (Directory.Exists(source))
+                {
+                    // If the source is a known directory, we can use the rest of the metadata to do a precice prediction.
```

Comment:
>precice [](start = 103, length = 7)

precise

Reply:
Plus anything else I mentioned in the internal review :)

---
In reply to: [301728848](https://github.com/microsoft/MSBuildPrediction/pull/35#discussion_r301728848) [](ancestors = 301728848)